### ISSUE TITLE
Ghidra error while trying to "Open PDB File" - Parameter type may not specify the void datatype 

### ISSUE THREAD

**Issue by DaniloNC** [NationalSecurityAgency/ghidra#2523]
**Describe the bug**
When trying to import function names from the PDB file I am receiving the error below.
This happens with both the PDB Universal and PDB MSDIA.

```
Parameter type may not specify the void datatype - empty parameter list should be used
java.lang.IllegalArgumentException: Parameter type may not specify the void datatype - empty parameter list should be used
	at ghidra.program.model.data.ParameterDefinitionImpl.validateDataType(ParameterDefinitionImpl.java:93)
	at ghidra.program.model.data.ParameterDefinitionImpl.<init>(ParameterDefinitionImpl.java:51)
	at ghidra.program.model.data.ParameterDefinitionImpl.<init>(ParameterDefinitionImpl.java:39)
	at ghidra.app.util.pdb.pdbapplicator.ArgumentsListTypeApplier.applyTo(ArgumentsListTypeApplier.java:144)
	at ghidra.app.util.pdb.pdbapplicator.AbstractFunctionTypeApplier.applyInternal(AbstractFunctionTypeApplier.java:192)
	at ghidra.app.util.pdb.pdbapplicator.AbstractFunctionTypeApplier.applyOrDeferForDependencies(AbstractFunctionTypeApplier.java:180)
	at ghidra.app.util.pdb.pdbapplicator.AbstractFunctionTypeApplier.applyFunction(AbstractFunctionTypeApplier.java:146)
	at ghidra.app.util.pdb.pdbapplicator.ProcedureTypeApplier.apply(ProcedureTypeApplier.java:71)
	at ghidra.app.util.pdb.pdbapplicator.PdbApplicator.processDataTypesSequentially(PdbApplicator.java:649)
	at ghidra.app.util.pdb.pdbapplicator.PdbApplicator.processSequentially(PdbApplicator.java:720)
	at ghidra.app.util.pdb.pdbapplicator.PdbApplicator.processTypes(PdbApplicator.java:235)
	at ghidra.app.util.pdb.pdbapplicator.PdbApplicator.applyTo(PdbApplicator.java:200)
	at pdb.LoadPdbTask.parseWithNewParser(LoadPdbTask.java:189)
	at pdb.LoadPdbTask$1.analysisWorkerCallback(LoadPdbTask.java:81)
	at ghidra.app.plugin.core.analysis.AutoAnalysisManager$AnalysisWorkerCommand.applyTo(AutoAnalysisManager.java:1700)
	at ghidra.app.plugin.core.analysis.AutoAnalysisManager$AnalysisTaskWrapper.run(AutoAnalysisManager.java:688)
	at ghidra.app.plugin.core.analysis.AutoAnalysisManager.startAnalysis(AutoAnalysisManager.java:788)
	at ghidra.app.plugin.core.analysis.AutoAnalysisManager.startAnalysis(AutoAnalysisManager.java:667)
	at ghidra.app.plugin.core.analysis.AutoAnalysisManager.startAnalysis(AutoAnalysisManager.java:632)
	at ghidra.app.plugin.core.analysis.AnalysisBackgroundCommand.applyTo(AnalysisBackgroundCommand.java:58)
	at ghidra.framework.plugintool.mgr.BackgroundCommandTask.run(BackgroundCommandTask.java:102)
	at ghidra.framework.plugintool.mgr.ToolTaskManager.run(ToolTaskManager.java:315)
	at java.base/java.lang.Thread.run(Thread.java:834)

---------------------------------------------------
Build Date: 2020-Nov-13 1112 EST
Ghidra Version: 9.2
Java Home: C:\Program Files\AdoptOpenJDK\jdk-11.0.9.101-hotspot
JVM Version: AdoptOpenJDK 11.0.9.1
OS: Windows 10 10.0 amd64
Workstation: 192.168.116.128
```

**To Reproduce**
1. Analyze the `echo_server_tlsd.exe`
2. Go to `File` -> `Load PDB File...`
3. Select the `echo_server_tlsd.pdb`

**Expected behavior**
Ghidra would import the PDB symbols file and have the correct function names

**Attachments**
[Target.zip](https://github.com/radareorg/radare2/files/5599564/target.zip)

**Environment (please complete the following information):**
- Build Date: 2020-Nov-13 1112 EST
- Ghidra Version: 9.2
- Java Home: C:\Program Files\AdoptOpenJDK\jdk-11.0.9.101-hotspot
- JVM Version: AdoptOpenJDK 11.0.9.1
- OS: Windows 10 10.0 amd64

**Comment by ghizard** [comment]
@DaniloNC Do you know what was used to compile this code and do you you know the underlying language?

We give the opinion of visualstudio:unknown for toolchain.

Microsoft's PDB format is supposed to be somewhat stable, though we have seen some data type record numbers that are not listed in their "documentation" (source code on github).

I found the source of the problem above:  a pointer's underlying type was unknown, and I was setting the pointer to void when I meant to set the underlying type as unknown.  I have that fix in progress.

I have run into another issue where I was strict in trying to set certain types.  In particular, I'm looking at the record of a member function, and this record holds the record number for the "containing class," which could be a class, structure (, union), and during development, I discovered enum (which I think is not valid for C++, but I think the source was Rust).  I try to handle "unknown" types, but in the case of your binary, the "containing class" is one of these unknown types.

On occasion, I'll find tidbits of information regarding PDB from other sources, such as LLVM documentation, but they don't have information on the types either.  Type ID of the "containing class" was 0x1608.  And the other unknown type in your binary has ID 0x1609, and it was the underlying type of a pointer (which was what caused the exception you gave above, where the pointer to this type was the first argument in the arguments list of a function).

If I find no other information, I'll try to analyze the data bytes for these record types, and I can put type 0x1608 into the acceptable "containing class" types, but it could be anything, so a hint of the underlying language could be helpful.  It could also just be a newer format for a class, structure, etc. that MSFT is using.

**Comment by DaniloNC** [comment]
Hi @ghizard, thanks for the follow-up.

The source code is at https://github.com/zaphoyd/websocketpp, this is a C++ header-only library that was compiled with the BUILD_EXAMPLES CMake flag. The dependencies (e.g. libboost, etc) were downloaded using the https://github.com/microsoft/vcpkg

Visual Studio 2019 was used to compile the project.

Let me know if you need a step-by-step guide on how I created the .exe/.pdb file and I can provide it to you later.

I was reversing I binary that uses the websocketpp library and I was trying to compile the project myself with symbols in order to create a new function ID database, hoping that this would identify some functions on my target binary.

**Comment by ghizard** [comment]
Ok.  Thanks.  Digging deeper, 0x1608 seems to be a new class type and 0x1609 a new struct type based on mangled names found in the records.  I hadn't done a lot of testing with VS 2019, but should hopefully be able to figure things out.

**Comment by ghizard** [comment]
The records seem almost identical to the 0x1504 "class" and 0x1505 "structure" records, except having eliminated a redundant "count" field (it is not needed, as a separate record contains the field elements).  Since 0x1506 and 0x1507 are union and enum, respectively, which have this same redundancy, my guess would be that there could be 0x160a and 0x160b would also be newer versions of these.

**Comment by ghizard** [comment]
This was fixed in the Ghidra 9.2.1 release.

Class and Structure types were created (they probably represent user class and structure "typedefs" but that is for further investigation). Fallback datatype logic was also fixed, which is what caused the void datatype issue.
